### PR TITLE
Fix the incorrect matching algorithm in ShapelessRecipes and ShapelessOreRecipe

### PR DESCRIPTION
**Pull request by ustc-zzzz**
The problem is found by @3TUSK .

* Firstly, create a recipe like this: <https://gist.github.com/3TUSK/e45ea99ad61ea67bdd803bf1cff1b350>
* Secondly, put these three items orderly in a crafting table: 
    * Nether Star (`minecraft:nether_star@0`)
    * Spruce Planks (`minecraft:planks@1`)
    * Birch Planks (`minecraft:planks@2`)
* You will find that nothing will be crafted while it should be
* Exchange the positions of Spruce Planks and Birch Planks, and the output stack appears

The reason is that, in the first pattern, Vanilla Minecraft firstly checks if the first item is `minecraft:nether_star@0` (it is), and secondly checks if the second item is `minecraft:planks@0`, `@1`, `@2`, `@3`, `@4`, or `@5` (certainly it is), and finally checks if the third item is `minecraft:planks@1`  (oh it's not). Then Vanilla Minecraft judges that it does not match the recipe.

~~In fact, if the third item does not matches the ingredient, it should try another way to match the second item. In other words, there should be a stack like this:~~
* ~~`[]` (initial state)~~
* ~~`[minecraft:nether_star@0]` (push the first item and it matches)~~
* ~~`[minecraft:nether_star@0, minecraft:planks@1]` (push the second item and it matches)~~
* ~~`[minecraft:nether_star@0, minecraft:planks@1, minecraft:planks@2]` (push the third item and it doesn't match)~~
* ~~`[minecraft:nether_star@0, minecraft:planks@1]` (pop the third item)~~
* ~~`[minecraft:nether_star@0, minecraft:planks@2]` (change the second item to another one)~~
* ~~`[minecraft:nether_star@0, minecraft:planks@2, minecraft:planks@1]` (it matches, good job)~~

~~The things shown above are what I did in the pull request.~~

`net.minecraft.client.util.RecipeItemHelper` has taken over everything.

### CODE REVIEW

**Comment by liach** [comment]
You should check sizes before since the loop is EXTREMELY slow. It uses O(n^3) time.

File: src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java [hunk 1]
```diff
@@ -90,38 +89,59 @@ public ShapelessOreRecipe(ResourceLocation group, @Nonnull ItemStack result, Obj
...
+            if (!itemstack.isEmpty())
             {
-                boolean inRecipe = false;
-                Iterator<Ingredient> req = required.iterator();
+                items.add(itemstack);
+            }
+        }
+
+        int size = items.size();
+        if (size > 0)
```
Comment (liach) [inline src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java]: `if (size == input.size())`

Comment (ustc-zzzz) [reply src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java]: @liach Fixed.

**Comment by liach** [comment]
Can you add a test mod that adds the recipe as you've described?

**Comment by mezz** [comment]
This seems to be a vanilla bug, please report it to mojang: https://bugs.mojang.com/projects/MC/issues
and then link it here.

This implementation needs some fixing up, because the patch is pretty large.
The preferred solution is probably to put the correct code in Forge somewhere, and then have a one-line hook at the top of the vanilla method which calls the Forge one. That keeps the patch small.

**Comment by liach** [comment]
Moreover, what is the exact time complexity of this solution performance-wise? I think we may use [bipartite graph maximum matching](https://www.google.com/search?q=bipartite+graph+maximum+matching) for pairing things.

**Comment by 3TUSK** [comment]
As the original founder (per @ustc-zzzz said), I think we should fix this for `getRemainingItems` too?

**Comment by liach** [comment]
@3TUSK For now, there will be no difference in remaining items (it is just getting container items for everything)

Edit: You mean "finder" not "founder", right?

**Comment by liach** [comment]
Actually, your piece of code is recursion in nature. I fear that it may potentially get too slow.

**Comment by ustc-zzzz** [comment]
@liach Vanilla Minecraft's algorithm is O(n^2), and I'm using [backtracking algorithm](https://en.wikipedia.org/wiki/Backtracking).
In most cases, it doesn't need to push and pop the stack repeatedly (maybe I should call it "backtracking and forwardtracking"?) when matching a recipe, and the time complexity is still O(n^2).
The situation requiring extra time only happens when there are duplicate items in different ingredients as is shown in the GitHub gist, and I think that these special cases are quite rare.
What's more, recursion is inevitable because it is based on backtracking algorithm. You said that it is equivalent to a matching problem in graph theory, but I cannot find out what are edges and vertices in this problem. Maybe providing more detailed instructions or some codes is a better choice for me to understand how to use a more efficient algorithm.

**Comment by liach** [comment]
@ustc-zzzz We can understand the problem like this: (I will write in Chinese)

First, we get a graph for all the matching options by looping over both ingredients and item stacks in `O(n^2)` time. Then, we can assume the matching to be a bipartite graph, using the most efficient way (probably [Hopcroft–Karp algorithm](https://en.wikipedia.org/wiki/Hopcroft%E2%80%93Karp_algorithm), of time complexity `O(E(V)^(1/2))`, about `O(n^2)`) to match them up perfectly. This way, we can get a paired up configuration, which may be eventually used for custom remaining items; the time efficiency is not lowered by too much, too. We can add tests in test code for this algorithm to make sure it is correct.

我们可以把合成配料和物品都遍历一遍，时间效率`O(n^2)`，求出所有可行的配对。然后我们可以作二分图，用算法解决（比如[Hopcroft Karp算法](https://zh.wikipedia.org/wiki/%E9%9C%8D%E6%99%AE%E5%85%8B%E6%B4%9B%E5%A4%AB%E7%89%B9%EF%BC%8D%E5%8D%A1%E6%99%AE%E7%AE%97%E6%B3%95)，时间效率`O(E(V)^(1/2))`，接近 `O(n^2)`）这样一来我们可以得到一组完全匹配，同时也可以避免回溯出现指数级别的时间效率。可以在测试源码中加测试样例保证此算法正确。

**Comment by grum** [comment]
It is broken in vanilla! Thanks for the report (eventhough it is on the wrong spot :D)

@LexManos

Made the forge-version too now :D

```diff
diff --git a/patches/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java.patch b/patches/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java.patch
index cc918d8fa..9875f5521 100644
--- a/patches/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java.patch
+++ b/patches/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java.patch
@@ -32,7 +32,48 @@
          }

          return nonnulllist;
-@@ -136,7 +130,6 @@
+@@ -61,7 +55,8 @@
+
+     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
+     {
+-        List<Ingredient> list = Lists.newArrayList(this.field_77579_b);
++        final net.minecraft.client.util.RecipeItemHelper helper = new net.minecraft.client.util.RecipeItemHelper();
++        int ingredientCount = 0;
+
+         for (int i = 0; i < p_77569_1_.func_174923_h(); ++i)
+         {
+@@ -71,27 +66,13 @@
+
+                 if (!itemstack.func_190926_b())
+                 {
+-                    boolean flag = false;
+-
+-                    for (Ingredient ingredient : list)
+-                    {
+-                        if (ingredient.apply(itemstack))
+-                        {
+-                            flag = true;
+-                            list.remove(ingredient);
+-                            break;
+-                        }
+-                    }
+-
+-                    if (!flag)
+-                    {
+-                        return false;
+-                    }
++                    ingredientCount++;
++                    helper.func_194112_a(itemstack);
+                 }
+             }
+         }
+
+-        return list.isEmpty();
++        return ingredientCount == this.field_77579_b.size() && helper.func_194116_a(this, null);
+     }
+
+     public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
+@@ -136,7 +117,6 @@
          return nonnulllist;
      }
```

```diff
diff --git a/Minecraft.Server/src/main/java/net/minecraft/world/item/crafting/ShapelessRecipe.java b/Minecraft.Server/src/main/java/net/minecraft/world/item/crafting/ShapelessRecipe.java
index 18fde0132c..4c8e94ee8d 100644
--- a/Minecraft.Server/src/main/java/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/Minecraft.Server/src/main/java/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -6,6 +6,7 @@ import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
 import net.minecraft.core.NonNullList;
 import net.minecraft.util.GsonHelper;
+import net.minecraft.world.entity.player.StackedContents;
 import net.minecraft.world.inventory.CraftingContainer;
 import net.minecraft.world.item.ItemInstance;
 import net.minecraft.world.level.Level;
@@ -54,29 +55,20 @@ public class ShapelessRecipe implements Recipe {
 
     @Override
     public boolean matches(final CraftingContainer craftSlots, final Level level) {
-        final List<Ingredient> unusedIngredients = Lists.newArrayList(ingredients);
+        final StackedContents stackedContents = new StackedContents();
+        int ingredientCount = 0;
 
         for (int y = 0; y < craftSlots.getHeight(); y++) {
             for (int x = 0; x < craftSlots.getWidth(); x++) {
                 final ItemInstance item = craftSlots.getItem(x, y);
-
                 if (!item.isEmpty()) {
-                    boolean found = false;
-                    for (final Ingredient ingredient : unusedIngredients) {
-                        if (ingredient.apply(item)) {
-                            found = true;
-                            unusedIngredients.remove(ingredient);
-                            break;
-                        }
-                    }
-                    if (!found) {
-                        return false;
-                    }
+                    stackedContents.accountStack(item);
+                    ingredientCount++;
                 }
             }
         }
 
-        return unusedIngredients.isEmpty();
+        return ingredientCount == ingredients.size() && stackedContents.canCraft(this, null);
     }
 
     @Override
```

**Comment by ustc-zzzz** [comment]
Some helpful information:
`net.minecraft.world.entity.player.StackedContents` => `net.minecraft.client.util.RecipeItemHelper`
`net.minecraft.world.inventory.CraftingContainer` => `net.minecraft.inventory.InventoryCrafting`

**Comment by grum** [comment]
Haha nice you turned it into a short Forge patch as well, \o/.

**Comment by liach** [comment]
Would this patch crash on server side? That helper thing seems to be client only right now. (only for forge, as vanilla can repackage)

File: src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java [hunk 2]
```diff
@@ -19,25 +19,25 @@
 
 package net.minecraftforge.oredict;
 
-import java.util.Iterator;
+import java.util.List;
 import net.minecraft.block.Block;
+import net.minecraft.client.util.RecipeItemHelper;
```
Comment (liach) [inline src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java:24]: Are you sure this is available on the server side? You might remove the side annotation, I guess.

Comment (ustc-zzzz) [reply src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java:24]: The class is available on both server side and client side.
In my opinion, mcp did something wrong with the package of the class.

**Review by liach — APPROVED** [review]

File: src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java (hunk 2 above)
Comment (LexManos) [reply src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java:24]: No I didn't. Mojang did by making it server side in the hotfix. I didn't feel like nuking all mods by renaming it on a small patch.

Comment (grum) [reply src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java:24]: Pfew! Then again it compiled and worked here :D

**Comment by LexManos** [comment]
Just as a update, this broke 'tool' items, as discussed here: https://github.com/MinecraftForge/MinecraftForge/issues/4516
That has been resolved by this commit: https://github.com/MinecraftForge/MinecraftForge/commit/89db87dbfce4e4bd130f5daccac7cac495002275

**Pull request merged**